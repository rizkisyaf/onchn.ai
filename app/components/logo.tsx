import { motion } from 'framer-motion'
import Image from 'next/image'

interface LogoProps {
  size?: number
  className?: string
}

export function Logo({ size = 40, className = '' }: LogoProps) {
  return (
    <motion.div 
      className={`relative ${className}`}
      style={{ width: size, height: size }}
      whileHover={{ scale: 1.1 }}
      whileTap={{ scale: 0.95 }}
    >
      <Image
        src="/images/logo/logo-mark.svg"
        alt="onchn.ai logo"
        width={size}
        height={size}
        className="w-full h-full"
      />
    </motion.div>
  )
} 