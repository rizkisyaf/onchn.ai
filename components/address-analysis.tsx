'use client'

import { useEffect } from 'react'
import { Card } from '@/components/ui/card'
import { Progress } from '@/components/ui/progress'
import { useAddressDetection } from '@/hooks/use-address-detection'
import { AddressMetadata } from '@/types/wallet'

interface AddressAnalysisProps {
  address: string
  onDetection?: (metadata: AddressMetadata) => void
}

export function AddressAnalysis({ address, onDetection }: AddressAnalysisProps) {
  const { detect, isDetecting, error } = useAddressDetection({
    onDetection,
  })

  useEffect(() => {
    if (address) {
      detect(address)
    }
  }, [address, detect])

  if (error) {
    return (
      <Card className="p-4">
        <div className="text-red-500">
          <h3 className="font-semibold">Error</h3>
          <p>{error.message}</p>
        </div>
      </Card>
    )
  }

  if (isDetecting) {
    return (
      <Card className="p-4">
        <div className="space-y-4">
          <h3 className="font-semibold">Detecting Address Type...</h3>
          <Progress value={50} />
        </div>
      </Card>
    )
  }

  return (
    <Card className="p-4">
      <div className="space-y-4">
        <h3 className="font-semibold">Address Analysis</h3>
        <p className="text-sm text-muted-foreground">
          Select an address to analyze its type and properties.
        </p>
      </div>
    </Card>
  )
} 