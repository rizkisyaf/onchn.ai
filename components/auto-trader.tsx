'use client'

import { useEffect, useState } from 'react'
import { motion } from 'framer-motion'
import { Card } from '@/components/ui/card'
import { Button } from '@/components/ui/button'
import { Input } from '@/components/ui/input'
import { Label } from '@/components/ui/label'
import { Switch } from '@/components/ui/switch'
import { ArrowUpDown, Wallet } from 'lucide-react'
import { useToast } from '@/components/ui/use-toast'
import { handleComponentError } from '@/lib/error-handling'

interface AutoTraderProps {
  isActive: boolean
}

type TradeType = 'buy' | 'sell'

interface Trade {
  pair: string
  time: Date
  amount: number
  type: TradeType
}

export function AutoTrader({ isActive }: AutoTraderProps) {
  const [balance, setBalance] = useState<number>(0)
  const [profit, setProfit] = useState<number>(0)
  const [trades, setTrades] = useState<number>(0)
  const [maxTradeSize, setMaxTradeSize] = useState<string>('')
  const [stopLoss, setStopLoss] = useState<string>('')
  const [isLoading, setIsLoading] = useState<boolean>(false)
  const [recentTrades, setRecentTrades] = useState<Trade[]>([])
  const { toast } = useToast()

  const validateSettings = () => {
    const errors: string[] = []
    
    if (maxTradeSize && (parseFloat(maxTradeSize) <= 0 || parseFloat(maxTradeSize) > 100)) {
      errors.push('Maximum trade size must be between 0 and 100 SOL')
    }
    
    if (stopLoss && (parseFloat(stopLoss) <= 0 || parseFloat(stopLoss) > 50)) {
      errors.push('Stop loss must be between 0 and 50%')
    }

    return errors
  }

  const handleMaxTradeSizeChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value
    if (value === '' || (/^\d*\.?\d*$/.test(value))) {
      setMaxTradeSize(value)
    }
  }

  const handleStopLossChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const value = e.target.value
    if (value === '' || (/^\d*\.?\d*$/.test(value))) {
      setStopLoss(value)
    }
  }

  useEffect(() => {
    if (isActive) {
      try {
        setIsLoading(true)
        const errors = validateSettings()
        
        if (errors.length > 0) {
          errors.forEach(error => {
            toast({
              title: 'Invalid Settings',
              description: error
            })
          })
          return
        }

        // Simulate trading activity
        const interval = setInterval(() => {
          setBalance(prev => +(prev + Math.random() * 0.1).toFixed(2))
          setProfit(prev => +(prev + Math.random() * 0.05).toFixed(2))
          setTrades(prev => prev + 1)
          
          const tradeType: TradeType = Math.random() > 0.5 ? 'buy' : 'sell'
          // Add new trade to recent trades
          setRecentTrades(prev => [{
            pair: 'SOL/USDC',
            time: new Date(),
            amount: +(Math.random() * 0.1).toFixed(2),
            type: tradeType
          }, ...prev].slice(0, 3))
        }, 5000)

        setIsLoading(false)
        return () => clearInterval(interval)
      } catch (error) {
        handleComponentError(error as Error, 'AutoTrader useEffect')
        toast({
          title: 'Error',
          description: 'Failed to start auto trader. Please try again.'
        })
        setIsLoading(false)
      }
    }
  }, [isActive, toast])

  return (
    <div className="space-y-6">
      {/* Stats */}
      <div className="grid grid-cols-1 md:grid-cols-3 gap-6">
        <motion.div
          initial={{ opacity: 0, y: 20 }}
          animate={{ opacity: 1, y: 0 }}
          transition={{ delay: 0.1 }}
        >
          <div className="bg-white/5 rounded-xl p-6 backdrop-blur-sm border border-white/10">
            <h3 className="text-sm font-medium text-gray-400 mb-2">Balance</h3>
            <p className="text-2xl font-bold">{balance.toFixed(2)} SOL</p>
          </div>
        </motion.div>

        <motion.div
          initial={{ opacity: 0, y: 20 }}
          animate={{ opacity: 1, y: 0 }}
          transition={{ delay: 0.2 }}
        >
          <div className="bg-white/5 rounded-xl p-6 backdrop-blur-sm border border-white/10">
            <h3 className="text-sm font-medium text-gray-400 mb-2">Profit/Loss</h3>
            <p className={`text-2xl font-bold ${profit >= 0 ? 'text-green-500' : 'text-red-500'}`}>
              {profit >= 0 ? '+' : ''}{profit.toFixed(2)} SOL
            </p>
          </div>
        </motion.div>

        <motion.div
          initial={{ opacity: 0, y: 20 }}
          animate={{ opacity: 1, y: 0 }}
          transition={{ delay: 0.3 }}
        >
          <div className="bg-white/5 rounded-xl p-6 backdrop-blur-sm border border-white/10">
            <h3 className="text-sm font-medium text-gray-400 mb-2">Total Trades</h3>
            <p className="text-2xl font-bold">{trades}</p>
          </div>
        </motion.div>
      </div>

      {/* Settings */}
      <motion.div
        initial={{ opacity: 0, y: 20 }}
        animate={{ opacity: 1, y: 0 }}
        transition={{ delay: 0.4 }}
      >
        <div className="bg-white/5 rounded-xl p-6 backdrop-blur-sm border border-white/10">
          <h3 className="text-xl font-semibold mb-4">Trading Settings</h3>
          <div className="space-y-6">
            <div className="grid grid-cols-1 md:grid-cols-2 gap-6">
              <div className="space-y-2">
                <Label>Maximum Trade Size</Label>
                <Input 
                  type="text"
                  placeholder="0.00"
                  className="bg-white/5 border-white/10"
                  value={maxTradeSize}
                  onChange={handleMaxTradeSizeChange}
                  onBlur={() => {
                    if (maxTradeSize && parseFloat(maxTradeSize) > 100) {
                      toast({
                        title: 'Warning',
                        description: 'Maximum trade size seems high. Please verify.'
                      })
                    }
                  }}
                />
              </div>
              <div className="space-y-2">
                <Label>Stop Loss (%)</Label>
                <Input 
                  type="text"
                  placeholder="0.00"
                  className="bg-white/5 border-white/10"
                  value={stopLoss}
                  onChange={handleStopLossChange}
                  onBlur={() => {
                    if (stopLoss && parseFloat(stopLoss) > 50) {
                      toast({
                        title: 'Warning',
                        description: 'Stop loss percentage is quite high. Consider a lower value.'
                      })
                    }
                  }}
                />
              </div>
            </div>
            <div className="flex items-center justify-between">
              <div className="space-y-1">
                <Label>Auto Compound Profits</Label>
                <p className="text-sm text-gray-400">Automatically reinvest profits into new trades</p>
              </div>
              <Switch />
            </div>
          </div>
        </div>
      </motion.div>

      {/* Recent Trades */}
      <motion.div
        initial={{ opacity: 0, y: 20 }}
        animate={{ opacity: 1, y: 0 }}
        transition={{ delay: 0.5 }}
      >
        <div className="bg-white/5 rounded-xl p-6 backdrop-blur-sm border border-white/10">
          <h3 className="text-xl font-semibold mb-4">Recent Trades</h3>
          <div className="space-y-4">
            {recentTrades.length === 0 ? (
              <p className="text-gray-400 text-center py-4">No trades yet</p>
            ) : (
              recentTrades.map((trade, i) => (
                <div key={i} className="flex items-center justify-between p-4 bg-white/5 rounded-lg">
                  <div className="flex items-center gap-4">
                    <ArrowUpDown className={trade.type === 'buy' ? "text-green-500" : "text-red-500"} />
                    <div>
                      <p className="font-medium">{trade.pair}</p>
                      <p className="text-sm text-gray-400">
                        {new Date(trade.time).toLocaleTimeString()}
                      </p>
                    </div>
                  </div>
                  <p className={trade.type === 'buy' ? "text-green-500" : "text-red-500"}>
                    {trade.type === 'buy' ? '+' : '-'}{trade.amount.toFixed(2)} SOL
                  </p>
                </div>
              ))
            )}
          </div>
        </div>
      </motion.div>
    </div>
  )
} 