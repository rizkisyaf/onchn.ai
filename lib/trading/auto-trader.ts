import { ConfirmedSignatureInfo, Connection, ParsedAccountData, PublicKey } from '@solana/web3.js'
import { JupiterClient } from './jupiter-client'
import { BehaviorModel, WalletState, TradeAction } from '../ai/behavior-model'
import { TOKEN_PROGRAM_ID } from '@solana/spl-token'
import { Token } from '@/types/wallet'

export interface TradeParams {
  walletAddress: string
  maxAmount: number
  slippage: number
}

export interface TradeResult {
  success: boolean
  txid?: string
  type?: TradeAction['type']
  token?: string
  amount?: number
  confidence?: number
  error?: string
}

export class AutoTrader {
  private jupiterClient: JupiterClient
  private behaviorModel: BehaviorModel
  private connection: Connection

  constructor(connection: Connection, userPublicKey: PublicKey) {
    this.connection = connection
    this.jupiterClient = new JupiterClient(connection, userPublicKey)
    this.behaviorModel = new BehaviorModel()
  }

  async init() {
    await this.jupiterClient.init()
  }

  async executeTradeStrategy(params: TradeParams): Promise<TradeResult> {
    try {
      const state = await this.getWalletState(params.walletAddress)
      const prediction = await this.behaviorModel.predict(state)

      if (prediction.confidence < 0.7) {
        return {
          success: false,
          error: 'Low confidence prediction',
        }
      }

      const routes = await this.jupiterClient.getRoutes({
        inputMint: prediction.token,
        outputMint: prediction.token,
        amount: Math.min(prediction.amount, params.maxAmount),
        slippage: params.slippage,
      })

      if (routes.length === 0) {
        return {
          success: false,
          error: 'No routes available for trade',
        }
      }

      const bestRoute = routes[0]
      const txid = await this.jupiterClient.executeSwap(bestRoute)

      return {
        success: true,
        txid,
        type: prediction.type,
        token: prediction.token,
        amount: prediction.amount,
        confidence: prediction.confidence,
      }
    } catch (error) {
      return {
        success: false,
        error: error instanceof Error ? error.message : 'Unknown error',
      }
    }
  }

  private async getWalletState(walletAddress: string): Promise<WalletState> {
    if (process.env.NODE_ENV === 'development') {
      return {
        transactionCount: 100,
        uniqueTokens: 5,
        avgTransactionValue: 1000,
        tradeFrequency: 0.5,
        profitRatio: 1.2,
        riskLevel: 0.3,
        timeInMarket: 180 * 24 * 60 * 60 * 1000,
        totalInflow: 5000,
        totalOutflow: 4000,
        lastActivity: Date.now(),
        tokens: [],
        stats: {
          totalValue: 1000,
          totalTransactions: 100,
          uniqueTokens: 5,
          avgTransactionValue: 1000,
          lastActivity: Date.now()
        }
      }
    }

    // Real production logic
    const transactions = await this.connection.getSignaturesForAddress(new PublicKey(walletAddress))
    const tokenAccounts = await this.connection.getParsedTokenAccountsByOwner(
      new PublicKey(walletAddress),
      { programId: TOKEN_PROGRAM_ID }
    )

    const state: WalletState = {
      transactionCount: transactions.length,
      uniqueTokens: tokenAccounts.value.length,
      avgTransactionValue: await this.calculateAvgTransactionValue(transactions),
      tradeFrequency: this.calculateTradeFrequency(transactions),
      profitRatio: await this.calculateProfitRatio(transactions),
      riskLevel: await this.calculateRiskLevel(transactions),
      timeInMarket: Date.now() - transactions[transactions.length - 1].blockTime! * 1000,
      totalInflow: await this.calculateTotalInflow(transactions),
      totalOutflow: await this.calculateTotalOutflow(transactions),
      lastActivity: transactions[0].blockTime! * 1000,
      tokens: await this.getTokenDetails(tokenAccounts),
      stats: {
        totalValue: await this.calculateTotalValue(tokenAccounts),
        totalTransactions: transactions.length,
        uniqueTokens: tokenAccounts.value.length,
        avgTransactionValue: await this.calculateAvgTransactionValue(transactions),
        lastActivity: transactions[0].blockTime! * 1000
      }
    }

    return state
  }

  private async calculateAvgTransactionValue(transactions: ConfirmedSignatureInfo[]): Promise<number> {
    const txDetails = await Promise.all(
      transactions.slice(0, 100).map(tx => this.connection.getTransaction(tx.signature))
    )
    const values = txDetails.map(tx => tx?.meta?.fee || 0)
    return values.reduce((sum, val) => sum + val, 0) / values.length
  }

  private calculateTradeFrequency(transactions: ConfirmedSignatureInfo[]): number {
    if (transactions.length < 2) return 0
    const timeSpan = transactions[0].blockTime! - transactions[transactions.length - 1].blockTime!
    return transactions.length / (timeSpan / (24 * 60 * 60))
  }

  private async calculateProfitRatio(transactions: ConfirmedSignatureInfo[]): Promise<number> {
    const txDetails = await Promise.all(
      transactions.slice(0, 100).map(tx => this.connection.getTransaction(tx.signature))
    )
    const inflow = txDetails.reduce((sum, tx) => sum + (tx?.meta?.preBalances[0] || 0), 0)
    const outflow = txDetails.reduce((sum, tx) => sum + (tx?.meta?.postBalances[0] || 0), 0)
    return outflow / (inflow || 1)
  }

  private async calculateRiskLevel(transactions: ConfirmedSignatureInfo[]): Promise<number> {
    const txDetails = await Promise.all(
      transactions.slice(0, 100).map(tx => this.connection.getTransaction(tx.signature))
    )
    const volatility = this.calculateVolatility(txDetails)
    const frequency = this.calculateTradeFrequency(transactions)
    return Math.min(volatility * frequency / 100, 1)
  }

  private async calculateTotalInflow(transactions: ConfirmedSignatureInfo[]): Promise<number> {
    const txDetails = await Promise.all(
      transactions.slice(0, 100).map(tx => this.connection.getTransaction(tx.signature))
    )
    return txDetails.reduce((sum, tx) => sum + (tx?.meta?.preBalances[0] || 0), 0)
  }

  private async calculateTotalOutflow(transactions: ConfirmedSignatureInfo[]): Promise<number> {
    const txDetails = await Promise.all(
      transactions.slice(0, 100).map(tx => this.connection.getTransaction(tx.signature))
    )
    return txDetails.reduce((sum, tx) => sum + (tx?.meta?.postBalances[0] || 0), 0)
  }

  private async getTokenDetails(tokenAccounts: ParsedAccountData[]): Promise<Token[]> {
    return Promise.all(
      tokenAccounts.map(async account => {
        const mintInfo = await this.connection.getParsedAccountInfo(account.data.parsed.info.mint)
        return {
          address: account.data.parsed.info.mint,
          symbol: mintInfo.value?.data.parsed.info.symbol || 'UNKNOWN',
          name: mintInfo.value?.data.parsed.info.name || 'Unknown Token',
          decimals: account.data.parsed.info.tokenAmount.decimals,
          balance: Number(account.data.parsed.info.tokenAmount.amount),
          value: 0, // To be calculated with price feed
          price: 0, // To be fetched from price feed
          change24h: 0 // To be fetched from price feed
        }
      })
    )
  }

  private async calculateTotalValue(tokenAccounts: ParsedAccountData[]): Promise<number> {
    const tokens = await this.getTokenDetails(tokenAccounts)
    return tokens.reduce((sum, token) => sum + token.value, 0)
  }
} 