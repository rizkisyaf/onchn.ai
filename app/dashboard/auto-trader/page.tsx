'use client'

import { useState } from 'react'
import { motion } from 'framer-motion'
import { AutoTrader } from '@/components/auto-trader'
import { Card } from '@/components/ui/card'
import { Button } from '@/components/ui/button'
import { useToast } from '@/components/ui/use-toast'
import { Bot as RobotIcon, Power, Loader2 } from 'lucide-react'
import { ErrorBoundary } from '@/components/error-boundary'
import { useSubscription } from '@/hooks/useSubscription'

export default function AutoTraderPage() {
  const { canAccessFeature, currentPlan } = useSubscription()
  const canAccessAutoTrader = canAccessFeature('automatedTrading')
  const [isActive, setIsActive] = useState(false)
  const [isLoading, setIsLoading] = useState(false)
  const { toast } = useToast()

  if (!canAccessAutoTrader) {
    return (
      <div className="p-6">
        <Card className="p-6 bg-white/5 border-white/10">
          <h2 className="text-xl font-semibold mb-4">Feature Restricted</h2>
          <p className="text-gray-400 mb-4">
            Auto trading is only available for Pro tier users. Upgrade your subscription to access automated trading features.
          </p>
          <Button>Upgrade Now</Button>
        </Card>
      </div>
    )
  }

  const handleToggle = async () => {
    try {
      setIsLoading(true)
      // Simulate API call delay
      await new Promise(resolve => setTimeout(resolve, 1000))
      
      setIsActive(!isActive)
      toast({
        title: isActive ? 'Auto Trader Deactivated' : 'Auto Trader Activated',
        description: isActive ? 'Trading bot has been stopped' : 'Trading bot is now running',
      })
    } catch (error) {
      toast({
        title: 'Error',
        description: 'Failed to toggle auto trader. Please try again.',
      })
    } finally {
      setIsLoading(false)
    }
  }

  return (
    <div className="space-y-8">
      {/* Header */}
      <div className="flex items-center gap-3">
        <RobotIcon className="h-8 w-8 text-primary" />
        <motion.h1 
          initial={{ opacity: 0, y: 20 }}
          animate={{ opacity: 1, y: 0 }}
          className="text-3xl font-bold"
        >
          Auto Trader
        </motion.h1>
      </div>

      {/* Control Panel */}
      <motion.div
        initial={{ opacity: 0, y: 20 }}
        animate={{ opacity: 1, y: 0 }}
        transition={{ delay: 0.1 }}
      >
        <Card className="p-6 bg-white/5 border-white/10">
          <div className="flex items-center justify-between">
            <div>
              <h2 className="text-xl font-semibold mb-2">Trading Bot</h2>
              <p className="text-gray-400">
                {isActive 
                  ? 'Bot is actively monitoring and trading based on behavior analysis'
                  : 'Start the bot to begin automated trading'}
              </p>
            </div>
            <Button 
              onClick={handleToggle}
              size="lg"
              disabled={isLoading}
              className={`${
                isActive 
                  ? 'bg-accent hover:bg-accent/90' 
                  : 'bg-primary hover:bg-primary/90'
              } text-white rounded-full px-8`}
            >
              {isLoading ? (
                <>
                  <Loader2 className="mr-2 h-5 w-5 animate-spin" />
                  {isActive ? 'Stopping...' : 'Starting...'}
                </>
              ) : (
                <>
                  <Power className="mr-2 h-5 w-5" />
                  {isActive ? 'Stop' : 'Start'}
                </>
              )}
            </Button>
          </div>
        </Card>
      </motion.div>

      {/* Trading Interface */}
      <motion.div
        initial={{ opacity: 0, y: 20 }}
        animate={{ opacity: 1, y: 0 }}
        transition={{ delay: 0.2 }}
      >
        <AutoTrader isActive={isActive} />
      </motion.div>
    </div>
  )
} 