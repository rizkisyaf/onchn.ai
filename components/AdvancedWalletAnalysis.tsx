import React from 'react'
import { Card } from '@/components/ui'
import { WalletData } from '@/types/wallet'
import { LineChart, Line, XAxis, YAxis, CartesianGrid, Tooltip, Legend, ResponsiveContainer, PieChart, Pie, Cell } from 'recharts'

interface AdvancedWalletAnalysisProps {
  walletData: WalletData
}

const COLORS = ['#0088FE', '#00C49F', '#FFBB28', '#FF8042', '#8884D8']

export const AdvancedWalletAnalysis: React.FC<AdvancedWalletAnalysisProps> = ({ walletData }) => {
  const tokenDistribution = walletData.tokens.map(token => ({
    name: token.symbol,
    value: token.value
  }))

  const totalValue = walletData.tokens.reduce((sum, token) => sum + token.value, 0)

  const calculateDiversificationScore = () => {
    const herfindahlIndex = walletData.tokens.reduce((sum, token) => {
      const marketShare = token.value / totalValue
      return sum + marketShare * marketShare
    }, 0)
    return (1 - herfindahlIndex) * 100
  }

  const diversificationScore = calculateDiversificationScore()

  return (
    <Card className="p-6">
      <h2 className="text-2xl font-bold mb-4">Advanced Wallet Analysis</h2>
      
      <div className="mb-6">
        <h3 className="text-xl font-semibold mb-2">Portfolio Composition</h3>
        <ResponsiveContainer width="100%" height={300}>
          <PieChart>
            <Pie
              data={tokenDistribution}
              dataKey="value"
              nameKey="name"
              cx="50%"
              cy="50%"
              outerRadius={80}
              fill="#8884d8"
              label
            >
              {tokenDistribution.map((entry, index) => (
                <Cell key={`cell-${index}`} fill={COLORS[index % COLORS.length]} />
              ))}
            </Pie>
            <Tooltip />
            <Legend />
          </PieChart>
        </ResponsiveContainer>
      </div>

      <div className="mb-6">
        <h3 className="text-xl font-semibold mb-2">Diversification Score</h3>
        <p className="text-3xl font-bold">{diversificationScore.toFixed(2)}%</p>
        <p className="text-gray-600">
          {diversificationScore > 70 ? 'Well diversified' : 'Consider diversifying your portfolio'}
        </p>
      </div>

      <div>
        <h3 className="text-xl font-semibold mb-2">Token Performance</h3>
        <ResponsiveContainer width="100%" height={300}>
          <LineChart data={walletData.tokens}>
            <CartesianGrid strokeDasharray="3 3" />
            <XAxis dataKey="symbol" />
            <YAxis />
            <Tooltip />
            <Legend />
            <Line type="monotone" dataKey="value" stroke="#8884d8" />
          </LineChart>
        </ResponsiveContainer>
      </div>
    </Card>
  )
}

