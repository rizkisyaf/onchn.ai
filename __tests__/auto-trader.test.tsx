import { render, screen } from '@testing-library/react'
import { AutoTrader } from '@/components/auto-trader'

describe('AutoTrader', () => {
  it('renders in inactive state', () => {
    render(<AutoTrader isActive={false} />)
    expect(screen.getByText(/Trading Settings/i)).toBeInTheDocument()
  })

  it('shows loading state when active', () => {
    const { container } = render(<AutoTrader isActive={true} />)
    expect(container).toMatchSnapshot()
  })

  it('updates when new trade occurs', () => {
    const { container: newContainer } = render(<AutoTrader isActive={true} />)
    expect(newContainer).toMatchSnapshot()
  })

  it('displays error state', () => {
    render(<AutoTrader isActive={false} />)
    expect(screen.getByText(/Trading Settings/i)).toBeInTheDocument()
  })

  it('handles trade execution', () => {
    render(<AutoTrader isActive={true} />)
    expect(screen.getByText(/Trading Settings/i)).toBeInTheDocument()
  })

  it('handles trade cancellation', () => {
    render(<AutoTrader isActive={false} />)
    expect(screen.getByText(/Trading Settings/i)).toBeInTheDocument()
  })
}) 