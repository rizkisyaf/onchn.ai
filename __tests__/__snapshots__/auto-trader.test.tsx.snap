// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AutoTrader shows loading state when active 1`] = `
<div>
  <div
    class="space-y-6"
  >
    <div
      class="grid grid-cols-1 md:grid-cols-3 gap-6"
    >
      <div
        style="opacity: 0; transform: translateY(20px);"
      >
        <div
          class="bg-white/5 rounded-xl p-6 backdrop-blur-sm border border-white/10"
        >
          <h3
            class="text-sm font-medium text-gray-400 mb-2"
          >
            Balance
          </h3>
          <p
            class="text-2xl font-bold"
          >
            0.00
             SOL
          </p>
        </div>
      </div>
      <div
        style="opacity: 0; transform: translateY(20px);"
      >
        <div
          class="bg-white/5 rounded-xl p-6 backdrop-blur-sm border border-white/10"
        >
          <h3
            class="text-sm font-medium text-gray-400 mb-2"
          >
            Profit/Loss
          </h3>
          <p
            class="text-2xl font-bold text-green-500"
          >
            +
            0.00
             SOL
          </p>
        </div>
      </div>
      <div
        style="opacity: 0; transform: translateY(20px);"
      >
        <div
          class="bg-white/5 rounded-xl p-6 backdrop-blur-sm border border-white/10"
        >
          <h3
            class="text-sm font-medium text-gray-400 mb-2"
          >
            Total Trades
          </h3>
          <p
            class="text-2xl font-bold"
          >
            0
          </p>
        </div>
      </div>
    </div>
    <div
      style="opacity: 0; transform: translateY(20px);"
    >
      <div
        class="bg-white/5 rounded-xl p-6 backdrop-blur-sm border border-white/10"
      >
        <h3
          class="text-xl font-semibold mb-4"
        >
          Trading Settings
        </h3>
        <div
          class="space-y-6"
        >
          <div
            class="grid grid-cols-1 md:grid-cols-2 gap-6"
          >
            <div
              class="space-y-2"
            >
              <label
                class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
              >
                Maximum Trade Size
              </label>
              <input
                class="flex h-10 w-full rounded-full border px-4 py-2 text-sm backdrop-blur-sm placeholder:text-gray-400 focus:outline-none focus:ring-2 focus:ring-primary/20 disabled:cursor-not-allowed disabled:opacity-50 bg-white/5 border-white/10"
                placeholder="0.00"
                type="text"
                value=""
              />
            </div>
            <div
              class="space-y-2"
            >
              <label
                class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
              >
                Stop Loss (%)
              </label>
              <input
                class="flex h-10 w-full rounded-full border px-4 py-2 text-sm backdrop-blur-sm placeholder:text-gray-400 focus:outline-none focus:ring-2 focus:ring-primary/20 disabled:cursor-not-allowed disabled:opacity-50 bg-white/5 border-white/10"
                placeholder="0.00"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="flex items-center justify-between"
          >
            <div
              class="space-y-1"
            >
              <label
                class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
              >
                Auto Compound Profits
              </label>
              <p
                class="text-sm text-gray-400"
              >
                Automatically reinvest profits into new trades
              </p>
            </div>
            <button
              aria-checked="false"
              class="peer inline-flex h-[24px] w-[44px] shrink-0 cursor-pointer items-center rounded-full border-2 border-transparent transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 focus-visible:ring-offset-background disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=unchecked]:bg-input"
              data-state="unchecked"
              role="switch"
              type="button"
              value="on"
            >
              <span
                class="pointer-events-none block h-5 w-5 rounded-full bg-background shadow-lg ring-0 transition-transform data-[state=checked]:translate-x-5 data-[state=unchecked]:translate-x-0"
                data-state="unchecked"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      style="opacity: 0; transform: translateY(20px);"
    >
      <div
        class="bg-white/5 rounded-xl p-6 backdrop-blur-sm border border-white/10"
      >
        <h3
          class="text-xl font-semibold mb-4"
        >
          Recent Trades
        </h3>
        <div
          class="space-y-4"
        >
          <p
            class="text-gray-400 text-center py-4"
          >
            No trades yet
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AutoTrader updates when new trade occurs 1`] = `
<div>
  <div
    class="space-y-6"
  >
    <div
      class="grid grid-cols-1 md:grid-cols-3 gap-6"
    >
      <div
        style="opacity: 0; transform: translateY(20px);"
      >
        <div
          class="bg-white/5 rounded-xl p-6 backdrop-blur-sm border border-white/10"
        >
          <h3
            class="text-sm font-medium text-gray-400 mb-2"
          >
            Balance
          </h3>
          <p
            class="text-2xl font-bold"
          >
            0.00
             SOL
          </p>
        </div>
      </div>
      <div
        style="opacity: 0; transform: translateY(20px);"
      >
        <div
          class="bg-white/5 rounded-xl p-6 backdrop-blur-sm border border-white/10"
        >
          <h3
            class="text-sm font-medium text-gray-400 mb-2"
          >
            Profit/Loss
          </h3>
          <p
            class="text-2xl font-bold text-green-500"
          >
            +
            0.00
             SOL
          </p>
        </div>
      </div>
      <div
        style="opacity: 0; transform: translateY(20px);"
      >
        <div
          class="bg-white/5 rounded-xl p-6 backdrop-blur-sm border border-white/10"
        >
          <h3
            class="text-sm font-medium text-gray-400 mb-2"
          >
            Total Trades
          </h3>
          <p
            class="text-2xl font-bold"
          >
            0
          </p>
        </div>
      </div>
    </div>
    <div
      style="opacity: 0; transform: translateY(20px);"
    >
      <div
        class="bg-white/5 rounded-xl p-6 backdrop-blur-sm border border-white/10"
      >
        <h3
          class="text-xl font-semibold mb-4"
        >
          Trading Settings
        </h3>
        <div
          class="space-y-6"
        >
          <div
            class="grid grid-cols-1 md:grid-cols-2 gap-6"
          >
            <div
              class="space-y-2"
            >
              <label
                class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
              >
                Maximum Trade Size
              </label>
              <input
                class="flex h-10 w-full rounded-full border px-4 py-2 text-sm backdrop-blur-sm placeholder:text-gray-400 focus:outline-none focus:ring-2 focus:ring-primary/20 disabled:cursor-not-allowed disabled:opacity-50 bg-white/5 border-white/10"
                placeholder="0.00"
                type="text"
                value=""
              />
            </div>
            <div
              class="space-y-2"
            >
              <label
                class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
              >
                Stop Loss (%)
              </label>
              <input
                class="flex h-10 w-full rounded-full border px-4 py-2 text-sm backdrop-blur-sm placeholder:text-gray-400 focus:outline-none focus:ring-2 focus:ring-primary/20 disabled:cursor-not-allowed disabled:opacity-50 bg-white/5 border-white/10"
                placeholder="0.00"
                type="text"
                value=""
              />
            </div>
          </div>
          <div
            class="flex items-center justify-between"
          >
            <div
              class="space-y-1"
            >
              <label
                class="text-sm font-medium leading-none peer-disabled:cursor-not-allowed peer-disabled:opacity-70"
              >
                Auto Compound Profits
              </label>
              <p
                class="text-sm text-gray-400"
              >
                Automatically reinvest profits into new trades
              </p>
            </div>
            <button
              aria-checked="false"
              class="peer inline-flex h-[24px] w-[44px] shrink-0 cursor-pointer items-center rounded-full border-2 border-transparent transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 focus-visible:ring-offset-background disabled:cursor-not-allowed disabled:opacity-50 data-[state=checked]:bg-primary data-[state=unchecked]:bg-input"
              data-state="unchecked"
              role="switch"
              type="button"
              value="on"
            >
              <span
                class="pointer-events-none block h-5 w-5 rounded-full bg-background shadow-lg ring-0 transition-transform data-[state=checked]:translate-x-5 data-[state=unchecked]:translate-x-0"
                data-state="unchecked"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      style="opacity: 0; transform: translateY(20px);"
    >
      <div
        class="bg-white/5 rounded-xl p-6 backdrop-blur-sm border border-white/10"
      >
        <h3
          class="text-xl font-semibold mb-4"
        >
          Recent Trades
        </h3>
        <div
          class="space-y-4"
        >
          <p
            class="text-gray-400 text-center py-4"
          >
            No trades yet
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
